src/agent/src/image_rpc.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs;
+use std::path::PathBuf;
+use std::process::{Command, ExitStatus};
+use std::sync::Arc;
+
+use anyhow::{anyhow, ensure, Result};
+use async_trait::async_trait;
+use protocols::image;
+use tokio::sync::Mutex;
+use ttrpc::{self, error::get_rpc_status as ttrpc_error};
+
+use crate::rpc::{verify_cid, CONTAINER_BASE};
+use crate::sandbox::Sandbox;
+
+const SKOPEO_PATH: &str = "/usr/bin/skopeo";
+const UMOCI_PATH: &str = "/usr/local/bin/umoci";
+const IMAGE_OCI: &str = "image_oci:latest";
+
+// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+pub struct ImageService {
+    sandbox: Arc<Mutex<Sandbox>>,
+}
+
+impl ImageService {
+    pub fn new(sandbox: Arc<Mutex<Sandbox>>) -> Self {
+        Self { sandbox }
+    }
+
+    fn build_oci_path(cid: &str) -> PathBuf {
+        let mut oci_path = PathBuf::from("/tmp");
+        oci_path.push(cid);
+        oci_path.push(IMAGE_OCI);
+        oci_path
+    }
+
+    fn pull_image_from_registry(image: &str, cid: &str) -> Result<()> {
+        let source_image = format!("{}{}", "docker://", image);
+
+        let mut manifest_path = PathBuf::from("/tmp");
+        manifest_path.push(cid);
+        manifest_path.push("image_manifest");
+        let target_path_manifest = format!("dir://{}", manifest_path.to_string_lossy());
+
+        // Define the target transport and path for the OCI image, without signature
+        let oci_path = Self::build_oci_path(cid);
+        let target_path_oci = format!("oci://{}", oci_path.to_string_lossy());
+
+        fs::create_dir_all(&manifest_path)?;
+        fs::create_dir_all(&oci_path)?;
+
+        info!(sl!(), "Attempting to pull image {}...", &source_image);
+
+        let mut pull_command = Command::new(SKOPEO_PATH);
+        pull_command
+            .arg("--insecure-policy")
+            .arg("copy")
+            .arg(source_image)
+            .arg(&target_path_manifest);
+
+        let status: ExitStatus = pull_command.status()?;
+        ensure!(
+            status.success(),
+            format!("failed to copy image manifest: {:?}", status)
```

Comment:
You don't need this `format!()`:

```suggestion
            "failed to copy image manifest: {:?}", status)
```

Same comment for the other `ensure!()` calls.

Reply:
fixed

Reply:
Thanks.